docs/source/advanced_ashpy.rst

```diff
@@ -0,0 +1,244 @@
+* :attr:`ashpy.metrics.metric.Metric.metric` (here referenced as :code:`self._metric` is the
+  `primitive` |keras.Metric| whose :code:`upadate_state()` method we will be using to simplify
+  our operations.
+* Custom computation will almost always be done via iteration over the data offered by the
+  |Context|.
+
+For a much more complex (but probably exhaustive) example have a look at the source code
+of :class:`ashpy.metrics.SlicedWassersteinDistance <ashpy.metrics.sliced_wasserstein_metric.SlicedWassersteinDistance>`.
+
+****************
+Custom Callbacks
+****************
+
+Our |ashpy.Callback| is built on the same base structure as a |keras.Callback| exposing methods
+acting as hooks for the same events.
+
+* on_train_start
+* on_epoch_start
+* on_batch_start
+* on_batch_end
+* on_epoch_end
+* on_train_end
+
+Inside the :mod:`ashpy.callbacks` module we offer two `primitive` Callbacks classes to inherit from.
+
+1. :class:`ashpy.callbacks.Callback <ashpy.callbacks.callback.Callback>`: is the most basic
+   form of callback and the basic block for all the other.
+2. |ashpy.CounterCallback|: is derived
+   from :class:`ashpy.callbacks.Callback <ashpy.callbacks.callback.Callback>` and contains
+   built-in logic for triggering an event given a desired frequency.
+
+Let's take a look at the following example which is the callback used to log GANs output to
+TensorBoard - :class:`ashpy.callbacks.gan.LogImageGANCallback`
+
+.. code-block:: python
+
+    class LogImageGANCallback(CounterCallback):
+        def __init__(
+            self,
+            event: Event = Event.ON_EPOCH_END,
+            name: str = "log_image_gan_callback",
+            event_freq: int = 1,
+        ) -> None:
+            """
+            Initialize the LogImageCallbackGAN.
+
+            Args:
+                event (:py:class:`ashpy.callbacks.events.Event`): event to consider.
+                event_freq (int): frequency of logging.
+                name (str): name of the callback.
+
+            """
+            super(LogImageGANCallback, self).__init__(
+                event=event, fn=self._log_fn, name=name, event_freq=event_freq
+            )
+
+        def _log_fn(self, context: GANContext) -> None:
+            """
+            Log output of the generator to Tensorboard.
+
+            Args:
+                context (:py:class:`ashpy.contexts.gan.GANContext`): current context.
+
+            """
+            if context.log_eval_mode == LogEvalMode.TEST:
+                out = context.generator_model(context.generator_inputs, training=False)
+            elif context.log_eval_mode == LogEvalMode.TRAIN:
+                out = context.fake_samples
+            else:
+                raise ValueError("Invalid LogEvalMode")
+
+            # tensorboard 2.0 does not support float images in [-1, 1]
+            # only in [0,1]
+            if out.dtype == tf.float32:
+                # The hypothesis is that image are in [-1,1] how to check?
```

Comment:
Maybe we can remove this

Reply:
I can remove it from the docs but should I remove it also from the code?

Reply:
No, I guess only for the doc. We should add a flag also for this callback to handle this.